Migrate to bitcoin::FeeRate

ValuedMammal opened the pull request:
### Description

This follows a similar approach to bitcoindevkit/bdk#1141 namely to remove `FeeRate` from `bdk::types` and instead defer to the upstream implementation for all fee rates. The idea is that making the switch allows BDK to benefit from a higher level abstraction, leaving the implementation details largely hidden.

As noted in bitcoindevkit/bdk#774, we should avoid extraneous conversions that can result in deviations in estimated transaction size and calculated fee amounts, etc. This would happen for example whenever calling a method like `FeeRate::to_sat_per_vb_ceil`. The only exception I would make is if we must return a fee rate error to the user, we might prefer to display it in the more familiar sats/vb, but this would only be useful if the rate can be expressed as a float.

### Notes to the reviewers

`bitcoin::FeeRate` is an integer whose native unit is sats per kilo-weight unit. In order to facilitate the change, a helper method `feerate_unchecked` is added and used only in wallet tests and psbt tests as necessary to convert existing fee rates to the new type. It's "unchecked" in the sense that we're not checking for integer overflow, because it's assumed we're passing a valid fee rate in a unit test.

Potential follow-ups can include:
- [x] Constructing a proper `FeeRate` from a `u64` in all unit tests, and thus obviating the need for the helper `feerate_unchecked` going forward.
- [x] Remove trait `Vbytes`.
- Consider adding an extra check that the argument to `TxBuilder::drain_to` is within "standard" size limits.
- Consider refactoring `coin_selection::select_sorted_utxos` to be efficient and readable.

closes bitcoindevkit/bdk#1136 

### Changelog notice

- Removed `FeeRate` type. All fee rates are now rust-bitcoin [`FeeRate`](https://docs.rs/bitcoin/latest/bitcoin/blockdata/fee_rate/struct.FeeRate.html).
- Removed trait `Vbytes`.

### Checklists

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

ValuedMammal committed:
refactor(bdk)!: drop FeeRate from bdk::types

Adopt `bitcoin::FeeRate` throughout

ValuedMammal committed:
test(bdk): initialize all feerates from `u64`

This makes the helper `feerate_unchecked` now redundant but
still usable.

ValuedMammal committed:
refactor(bdk): display CreateTxError::FeeRateTooLow in sat/vb

Also modify a unit test `test_bump_fee_low_fee_rate` to
additionally assert the expected error message

ValuedMammal committed:
chore: organize some imports

ValuedMammal commented:
After looking at some [related work](https://github.com/rust-bitcoin/rust-bitcoin/pull/2064#issuecomment-1736297479), I'm not totally opposed to displaying only whole numbers of fee rates. The difference between 99 and 100 sat/vb seems less significant than the difference between 1 and 2 sat/vb.

Some other conveniences I think would be handy:
- Saturating addition/subtraction of `FeeRate`s
- A sanity check that a given rate when applied to a minimum valid tx size doesn't produce an amount greater than `MAX_MONEY`

ValuedMammal commented on crates/bdk/src/wallet/mod.rs:
```diff
@@ -2385,6 +2385,17 @@ fn create_signers<E: IntoWalletDescriptor>(
+        $rate.to_sat_per_kwu() as f64 / ((1000 / WITNESS_SCALE_FACTOR) as f64)
+    }};
+}
+
```
suggest change doc to

```rust
/// Transforms a [`FeeRate`] to `f64` with unit of sat/vb
```

notmandatory commented:
>     * A sanity check that a given rate when applied to a minimum valid tx size doesn't produce an amount greater than `MAX_MONEY`

The fee and fee rate sanity check should be done in a new PR, I created bitcoindevkit/bdk_wallet#120 for this issue.

ValuedMammal commented:
Rebased to 553bdf92

- Resolved conflict with bitcoindevkit/bdk#1158 
- Fixed MSRV compile error

evanlinjin commented:
I think I have a controversial opinion about these changes. Because we plan to redo the entire tx building logic (after 1.0), and the current tx builder works okay as it is... Would this be worth the effort?

notmandatory commented:
This one needs a rebase after alpha.5 release.

ValuedMammal commented:
> I think I have a controversial opinion about these changes. Because we plan to redo the entire tx building logic (after 1.0), and the current tx builder works okay as it is... Would this be worth the effort?

I think it's fair to have a deeper discussion about this. I don't have strong feelings either way. I think I lean toward making the change because 1) ease of maintainability by aligning with rust bitcoin, 2) the units of a feerate can always be converted to any unit you like.

Granted I'm not as familiar with some of the long-standing design efforts of BDK, so I'm happy to get other opinions or hold off until everyone's in agreement.

storopoli commented on crates/bdk/src/wallet/coin_selection.rs:
```diff
@@ -313,7 +313,7 @@ impl CoinSelectionAlgorithm for OldestFirstCoinSelection {
     // drain_output_len = size(len(script_pubkey)) + len(script_pubkey) + size(output_value)
     let drain_output_len = serialize(drain_script).len() + 8usize;
-    let change_fee = fee_rate.fee_vb(drain_output_len);
+    let change_fee = (fee_rate * Weight::from_vb_unchecked(drain_output_len as u64)).to_sat();
```
I think we should use [`from_vb`](https://docs.rs/bitcoin/latest/bitcoin/blockdata/weight/struct.Weight.html#method.from_vb) here instead of `from_vb_unchecked`.

`from_vb` will do additional checks for overflow and returns an `Option` that we can put a nice error message in `.expect()`.

ValuedMammal replied:
I agree this is a concern. My opinion is that an insanely sized `drain_script` would overflow the block size before overflowing a `u64`, so this check should ideally occur during tx building and throw an error if the drain script exceeds either `MAX_STANDARD_TX_WEIGHT` (400kwu) or more generously `MAX_BLOCK_WEIGHT` (4Mwu). Note that if the drain script is derived from a typical bitcoin address, then this should never happen. I think adding these sanity checks early in tx building would make the code more robust, instead of cramming the logic into this particular line. What would be your take on that?

I would apply the same reasoning to `size_of_change` in the constructor of `BranchAndBoundCoinSelection`, but from what I hear, the coin select module is soon to be replaced so I don't know how relevant this is now.

In any case, your suggestion is sound, so we can either fixup this line of code, or I can add a note in the PR description about adding these extra consensus/standardness checks as a followup.

ValuedMammal replied:
I may be overthinking. We can do both - check for overflow here, and make a note to check the drain script size elsewhere

storopoli replied:
I don't think we need to check for overflow, in an error handling situation.
Just an `.expect()` with a nice error message would be sufficient in my opinion.
It will give enough context for the user of the source of the error.

storopoli commented on crates/bdk/src/wallet/coin_selection.rs:
```diff
@@ -456,7 +460,7 @@ impl CoinSelectionAlgorithm for BranchAndBoundCoinSelection {
             .fold(0, |acc, x| acc + x.effective_value);
 
-        let cost_of_change = self.size_of_change as f32 * fee_rate.as_sat_per_vb();
+        let cost_of_change = (Weight::from_vb_unchecked(self.size_of_change) * fee_rate).to_sat();
```
Same as the other comment, `from_vb`

storopoli commented on crates/bdk/src/wallet/coin_selection.rs:
```diff
@@ -893,7 +897,7 @@ mod test {
                 utxos,
                 vec![],
-                FeeRate::from_sat_per_vb(1.0),
+                FeeRate::from_sat_per_vb_unchecked(1),
```
These are fine to leave `*vb_unchecked` since they are tests

storopoli commented on crates/bdk/src/wallet/mod.rs:
```diff
@@ -1651,7 +1648,7 @@ impl<D> Wallet<D> {
     ///     let mut builder = wallet.build_fee_bump(tx.txid())?;
     ///     builder
-    ///         .fee_rate(bdk::FeeRate::from_sat_per_vb(5.0));
+    ///         .fee_rate(FeeRate::from_sat_per_vb_unchecked(5));
```
This is an example. I think we should use the more strict `from_sat_per_vb` to check for overflow.

Users will adhere to examples as the good practices for implementation.

ValuedMammal replied:
will fix

storopoli commented on crates/bdk/src/wallet/mod.rs:
```diff
@@ -2563,6 +2560,17 @@ fn create_signers<E: IntoWalletDescriptor>(
     Ok((signers, change_signers))
 }
 
+/// Transforms a [`FeeRate`] to `f64` with unit of sat/vb.
```
```suggestion
/// Transforms a [`FeeRate`] to `f64` with unit as sat/vb.
```

storopoli commented on crates/bdk/src/wallet/tx_builder.rs:
```diff
@@ -31,7 +31,7 @@
 //!     .add_recipient(to_address.script_pubkey(), 50_000)
 //!     // With a custom fee rate of 5.0 satoshi/vbyte
-//!     .fee_rate(bdk::FeeRate::from_sat_per_vb(5.0))
+//!     .fee_rate(FeeRate::from_sat_per_vb_unchecked(5))
```
This is an example. I think we should use the more strict `from_sat_per_vb` to check for overflow.

Users will adhere to examples as the good practices for implementation.

storopoli commented on crates/bdk/tests/psbt.rs:
```diff
@@ -10,6 +11,12 @@ use common::*;
+    let sat_kwu = (sat_vb * 250.0).ceil() as u64;
+    FeeRate::from_sat_per_kwu(sat_kwu)
+}
+
```
I might be missing something but I don't think you are using this function.

ValuedMammal replied:
This was kept in case someone in the future feels the need to construct a `bitcoin::FeeRate` from `f64`, since `bitcoin::FeeRate::from_sat_per_vb` only accepts an integer as param. We can also remove it now if you think it won't be missed.

storopoli replied:
No, this is good. It's in the test suite.
Let's add some docstrings then what you think?

storopoli commented on crates/bdk/tests/wallet.rs:
```diff
@@ -55,6 +55,12 @@ fn receive_output_in_latest_block(wallet: &mut Wallet, value: u64) -> OutPoint {
+    let sat_kwu = (sat_vb * 250.0).ceil() as u64;
+    FeeRate::from_sat_per_kwu(sat_kwu)
+}
+
```
I might be missing something but I don't think you are using this function.

storopoli replied:
Let's add some docstrings here as well, what you think?

storopoli commented on crates/bdk/tests/wallet.rs:
```diff
@@ -753,7 +765,7 @@ fn test_create_tx_drain_to_dust_amount() {
         .drain_to(addr.script_pubkey())
         .drain_wallet()
-        .fee_rate(FeeRate::from_sat_per_vb(453.0));
+        .fee_rate(FeeRate::from_sat_per_vb_unchecked(499));
```
Why changing from 453 to 499?

ValuedMammal replied:
Admittedly, this was a hack to make the test pass, and `499` was the first value chosen because it gives the impression of a "very high fee rate." The original `453` must have been right on the edge of passing. Evidently `454` would make this pass as well.

storopoli replied:
I can confirm that 454 passes the tests, can we use that?

```suggestion
        .fee_rate(FeeRate::from_sat_per_vb_unchecked(454));
```

storopoli commented on crates/bdk/tests/wallet.rs:
```diff
@@ -1568,8 +1590,9 @@ fn test_bump_fee_reduce_change() {
         .insert_tx(tx, ConfirmationTime::Unconfirmed { last_seen: 0 })
         .unwrap();
 
+    let feerate = FeeRate::from_sat_per_kwu(625); // 2.5 sat/vb
```
The math is not matching for me.
Am I doing something wrong?

`dbg!(psbt.clone().extract_tx().weight().to_wu() as f64 * 2.5);`

returns `1140.0`

ValuedMammal replied:
It looks like you're multiplying the tx weight by 2.5, which doesn't result in a feerate. To arrive at units of sat/kwu, you could do something like

`fee_amount * 1000 / tx.weight().to_wu()`

This will be more accurate if you `extract_tx` only after it has been signed.

storopoli replied:
Thanks!

storopoli reviewed (changes requested):
Some changes proposed.
Mostly important is to use [`from_vb`](https://docs.rs/bitcoin/latest/bitcoin/blockdata/weight/struct.Weight.html#method.from_vb) instead of `from_vb_unchecked` in code and examples.

Tests are fine to use the unchecked version.

ValuedMammal commented:
Thank you @storopoli , many good points brought up. I'm working through some of the comments now

storopoli commented on crates/bdk/src/types.rs:
```diff
@@ -247,70 +149,5 @@ impl Utxo {
-        assert!((fee.as_sat_per_vb() - 1.0).abs() < f32::EPSILON);
-        assert_eq!(fee.sat_per_kwu(), 250.0);
-    }
+    //use super::*;
```
I don't think this adds anything. Let's remove it instead of commenting?

ValuedMammal replied:
noted

storopoli reviewed (changes requested):
Just a minor thing.

ValuedMammal commented:
Recent changes:

- Using checked vsize-to-weight conversions in documentation and during coin selection.
- Added docstring to the test helper `feerate_unchecked` to explain its purpose, and moved it to bdk/tests/common.rs

evanlinjin commented on crates/bdk/src/wallet/coin_selection.rs:
```diff
@@ -344,9 +345,12 @@ fn select_sorted_utxos(
             (&mut selected_amount, &mut fee_amount),
```
I have a few thoughts here (I know this was like this before this PR, but I think we should address them).

1. Use a `for` loop instead of `.scan`. `.scan` makes this hard to read.
2. Because multiplying weight with feerate rounds up, we should keep an aggregate selection weight (instead of an aggregate `fee_amount`) and calculate the `fee_amount` on every iteration with the aggregate selection weight. This will give a more accurate result. Ref: https://docs.rs/bitcoin/latest/src/bitcoin/blockdata/fee_rate.rs.html#137-143.
3. This method needs doc comments. This methods requires that the `utxos` input is sorted by must-be-used UTXOs first otherwise it won't work as expected.

For 2., it is to be noted that what we have in this PR is already a lot better than what we had before (since our old `FeeRate::fee_wu` was a lot more aggressive when rounding up than the `FeeRate as Mul<Weight>` impl).

---

Edit: Ignore this! This PR is strictly better than what we had before. Also for point no. 2, we can't do the same for `OutputGroup` anyway so we shouldn't bother ourselves here.

In the future, we should focus our efforts on implementing our new `TxBuilder` than maintaining this old code that is to be scrapped.

ValuedMammal replied:
I agree this is a bit difficult to make sense of.

evanlinjin commented on crates/bdk/src/wallet/coin_selection.rs:
```diff
@@ -387,9 +391,10 @@ struct OutputGroup {
-            (TXIN_BASE_WEIGHT + weighted_utxo.satisfaction_weight) as u64,
-        ));
+        let fee = (fee_rate
+            * Weight::from_wu((TXIN_BASE_WEIGHT + weighted_utxo.satisfaction_weight) as u64))
```
~~Nit: Since we are doing this in multiple places, I think we should create a method `WeightedUtxo::txin_weight` that returns `TXIN_BASE_WEIGHT + self.satisfaction_weight`.~~

Edit: Ignore this also.

evanlinjin commented on crates/bdk/src/types.rs:
We can also get rid of `Vbytes` now since it is only used in one test. Yay!

evanlinjin reviewed (commented):
Thanks for working on this, fee/weight calculations should be more accurate than before since the old `FeeRate::fee_wu` was rounding up virtual bytes to whole units, whereas the `Mul<FeeRate> for Weight` impl (which we are using now) is rounding up kilo-wus.

However, I think we should try our best to avoid these changes in the future. This took a long time to review for me as I had to revisit old code to understand the ramifications of these changes. If there is no bug, we should just leave it as is so we can focus our efforts on the new `TxBuilder`.

As a final nit, let's get rid of `Vbytes` and I will happily ACK this!

ValuedMammal committed:
refactor(bdk)!: Remove trait Vbytes

The only place this is being used is a unit test that is
easily refactored. For size conversions prefer methods
on e.g. `Weight`.

ValuedMammal commented:
- Rebased onto d77a7f2ff13419d394e6a19f9942344a7c19a724
- Removed trait Vbytes in 011841905822a4ce58469bf1128cd018c85ce6d8

ValuedMammal commented on crates/bdk/tests/common.rs:
```diff
@@ -154,3 +154,16 @@ pub fn get_test_tr_with_taptree_xprv() -> &'static str {
+/// **Note** this 'quick and dirty' conversion should only be used when the input
+/// parameter has units of `satoshis/vbyte` **AND** is not expected to overflow,
+/// or else the resulting value will be inaccurate.
+fn feerate_unchecked(sat_vb: f64) -> FeeRate {
```
I guess this should be made `pub` for it to be at all useful.

ValuedMammal commented:
> Posting clippy checks here instead.
error: file opened with `create`, but `truncate` behavior not defined
  --> crates/file_store/src/store.rs:64:14
   |
64 |             .create(true)
   |              ^^^^^^^^^^^^- help: add: `.truncate(true)`
   |
   = help: if you intend to overwrite an existing file entirely, call `.truncate(true)`

@evanlinjin ?

evanlinjin commented:
@ValuedMammal rebase on master should fix it now!

evanlinjin reviewed (approved):
ACK 475a77219a6e7649c91b7cfeb44871cee9d0f374
